### PR TITLE
Add support for Subnet, and Network for GCP MachinePools

### PR DESCRIPTION
**Pull request by nimrodshn**
cc: @igoihman  @dgoodwin  @staebler  @gregsheremeta

### CODE REVIEW

File: pkg/apis/hive/v1/gcp/machinepools.go [hunk 1]
```diff
@@ -9,6 +9,20 @@ type MachinePool struct {
...
 
+	// ControlPlaneSubnet defines the subnet where the control plane will be deployed.
+	// The value should be the name of the subnet.
+	// If ommitted one will be created by the installer.
+	ControlPlaneSubnet string `json:"control_plane_subnet,omitempty"`
+
+	// ComputeSubnet defines the subnet where the compute nodes will be deployed.
+	// The value should be the name of the subnet.
+	// If ommitted one will be created by the installer.
+	ComputeSubnet string `json:"compute_subnet,omitempty"`
```
Comment (abhinavdahiya) [inline pkg/apis/hive/v1/gcp/machinepools.go:20]: What is the goal of letting the user define the control plane and compute subnet separately? seems like the user should only care which subnet this machinepool should be deployed..

From what i can see MachinePools today are mostly used to manage the `compute` role in cluster, (infra is still compute with additional stipulation) ... So maybe we treat the subnet as installertypes.platform.gcp.computeSubnet .

File: pkg/apis/hive/v1/gcp/machinepools.go [hunk 2]
```diff
@@ -9,6 +9,20 @@ type MachinePool struct {
 	// eg. n1-standard-4
 	InstanceType string `json:"type"`
 
+	// ControlPlaneSubnet defines the subnet where the control plane will be deployed.
+	// The value should be the name of the subnet.
+	// If ommitted one will be created by the installer.
```
Comment (abhinavdahiya) [inline pkg/apis/hive/v1/gcp/machinepools.go:14]: the doc comment seems not valid in the context of hive.

File: pkg/apis/hive/v1/gcp/machinepools.go [hunk 3]
```diff
@@ -9,6 +9,20 @@ type MachinePool struct {
...
+	ControlPlaneSubnet string `json:"control_plane_subnet,omitempty"`
+
+	// ComputeSubnet defines the subnet where the compute nodes will be deployed.
+	// The value should be the name of the subnet.
+	// If ommitted one will be created by the installer.
```
Comment (abhinavdahiya) [inline pkg/apis/hive/v1/gcp/machinepools.go:19]: the doc comment seems not valid in the context of hive.

File: pkg/apis/hive/v1/gcp/machinepools.go [hunk 4]
```diff
@@ -9,6 +9,20 @@ type MachinePool struct {
...
+	ComputeSubnet string `json:"compute_subnet,omitempty"`
+
+	// Network specifies an existing VPC where the cluster should be created
+	// rather than provisioning a new one.
+	Network string `json:"network,omitempty"`
```
Comment (abhinavdahiya) [inline pkg/apis/hive/v1/gcp/machinepools.go:24]: we must enforce that user set `network` and `subnet` both together or neither.

File: pkg/controller/remotemachineset/gcpactuator_test.go [hunk 5]
```diff
@@ -141,6 +141,48 @@ func TestGCPActuator(t *testing.T) {
...
+				pool := testGCPPool(testPoolName)
+				pool.Spec.Platform.GCP.Network = "some-vpc"
+				return pool
+			}(),
+			mockGCPClient: func(client *mockgcp.MockClient) {
+				mockListComputeZones(client, []string{"zone1"}, testRegion)
+			},
+			expectedMachineSetReplicas: map[string]int64{
+				generateGCPMachineSetName("worker", "zone1"): 3,
+			},
```
Comment (abhinavdahiya) [inline pkg/controller/remotemachineset/gcpactuator_test.go:184]: this is not a valid accepted case, setting a network with the subnet means hive can't know where to place the machine.

**Review by abhinavdahiya — CHANGES_REQUESTED** [review]
I tried to explain why allowing users to set the subnet on day-2 machinepools for GCP is prone to user errors here.

https://issues.redhat.com/browse/HIVE-1401?focusedCommentId=15873830&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-15873830
```
> https://github.com/kubernetes/kubernetes/blob/f6152d152140f42815765c7d9703b9817943c345/staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go#L99-L104

If you look at the code for the k8s-cloud-provider for gcp that handles the internal_load_balancer, you will notice that it only supports ALL nodes in the

SAME subnet.

We carry a patch in OpenShift https://github.com/kubernetes/kubernetes/pull/84466 that allows OpenShift to run control plane nodes in a separate subnet,

and all the rest in the same subnet. So all nodes, compute/infra/ or other that have different meaning as nodes in cluster must continue to be in the same

subnet in GCP context.

So I'm concerned that allowing users to provide subnet for day-2 MachinePools can create problems where if those nodes join the cluster, reconcilation for

all internal_load_balancers will be broken.

 

Installer allows setting only one pool of compute machines at install time and therefore the users cannot end up in this multi subnet situation, but hive MachinePools with subnet option has high chances of user error.
```

I can think of 2 ways to move forward, detailed in https://issues.redhat.com/browse/HIVE-1401?focusedCommentId=15873853&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-15873853
```
Can we get logs from the hive-controllers where BYO VPC was in use for GCP. The logs for remotemachineset controller should help understand what's broken.

Secondly, for now I would suggest that

A.

Hive use the already existing compute MachineSets in the cluster to configure the subnets/vpc for all the next MachinePools.

 

B.
If Hive supports the subnet on MachinePool, we enforce in validation webhooks that ALL MachinePools for a clusterdeployment use the same subnet.
```

File: pkg/apis/hive/v1/gcp/machinepools.go (hunk 1 above)
Comment (nimrodshn) [reply pkg/apis/hive/v1/gcp/machinepools.go:20]: > What is the goal of letting the user define the control plane and compute subnet separately? seems like the user should only care which subnet this machinepool should be deployed..

I was just exposing the API to the user as I saw the installer allowed for it. Alas, as you've mentioned - this _currently_ doesn't have a use case

> From what i can see MachinePools today are mostly used to manage the compute role in cluster, (infra is still compute with additional stipulation) ... So maybe we treat the subnet as installertypes.platform.gcp.computeSubnet .

So leave out `ControlPlaneSubnet` :+1:

File: pkg/apis/hive/v1/gcp/machinepools.go (hunk 4 above)
Comment (nimrodshn) [reply pkg/apis/hive/v1/gcp/machinepools.go:24]: Ack :+1:

File: pkg/controller/remotemachineset/gcpactuator_test.go (hunk 5 above)
Comment (nimrodshn) [reply pkg/controller/remotemachineset/gcpactuator_test.go:184]: @abhinavdahiya Did you mean a network _without_ a subnet? Will fix the test case.

File: pkg/apis/hive/v1/gcp/machinepools.go (hunk 3 above)
Comment (nimrodshn) [reply pkg/apis/hive/v1/gcp/machinepools.go:19]: Should I remove the last sentence? what would you suggest?

File: pkg/apis/hive/v1/gcp/machinepools.go [hunk 6]
```diff
@@ -9,6 +9,14 @@ type MachinePool struct {
 	// eg. n1-standard-4
 	InstanceType string `json:"type"`
 
+	// ComputeSubnet defines the subnet where the compute nodes will be deployed.
+	// The value should be the name of the subnet.
+	ComputeSubnet string `json:"compute_subnet,omitempty"`
```
Comment (abhinavdahiya) [inline pkg/apis/hive/v1/gcp/machinepools.go:14]:
```suggestion
	Subnet string `json:"subnet,omitempty"`
```

Comment (nimrodshn) [reply pkg/apis/hive/v1/gcp/machinepools.go:14]: Ack, thanks!

**Comment by nimrodshn** [comment]
@abhinavdahiya @dgoodwin With regards to you're comment:
> If Hive supports the subnet on MachinePool, we enforce in validation webhooks that ALL MachinePools for a clusterdeployment use the same subnet.

What if instead of validation at the MachinePool level we instead move these fields (i.e. `Network` and `ComputeSubnet` ) to the ClusterDeployment `gcp.Platform` field with the semantics being:
```
 This is the subnet where ALL compute machinepools will reside in
```
That way the invariant (all compute machinepools reside in the same subnet) is kept by the Actuator instead of the webhook validation; thus eliminating room for user errors

File: pkg/controller/remotemachineset/gcpactuator.go [hunk 7]
```diff
@@ -451,6 +463,32 @@ func getGCPImageID(masterMachine *machineapi.Machine, scheme *runtime.Scheme, lo
...
+	}
+
+	remoteMachineProviderSpec, err := decodeGCPMachineProviderSpec(
+		remoteMachineSets[0].Spec.Template.Spec.ProviderSpec.Value,
+		scheme,
+	)
+	if err != nil {
+		logger.WithError(err).Warn("cannot decode GCPMachineProviderSpec from master machine")
+		err = errors.Wrap(err, "cannot decode GCPMachineProviderSpec from master machine")
+		return
```
Comment (abhinavdahiya) [inline pkg/controller/remotemachineset/gcpactuator.go:480]:
```suggestion
		return "", "", errors.Wrap(err, "cannot decode GCPMachineProviderSpec from master machine")
```

using implicit returns creates confusion when reading the code as it doesn't become clear immediately that we are returning values compared to empty return..

File: pkg/controller/remotemachineset/gcpactuator.go [hunk 8]
```diff
@@ -451,6 +463,32 @@ func getGCPImageID(masterMachine *machineapi.Machine, scheme *runtime.Scheme, lo
...
+	}
+	if len(remoteMachineProviderSpec.NetworkInterfaces) == 0 {
+		logger.Warn("remove machine does not have any netowrk interfaces")
+		return
```
Comment (abhinavdahiya) [inline pkg/controller/remotemachineset/gcpactuator.go:484]: same return explicit error and empty values here.

File: pkg/controller/remotemachineset/gcpactuator.go [hunk 9]
```diff
@@ -451,6 +463,32 @@ func getGCPImageID(masterMachine *machineapi.Machine, scheme *runtime.Scheme, lo
...
+	}
+
+	network = remoteMachineProviderSpec.NetworkInterfaces[0].Network
+	subnet = remoteMachineProviderSpec.NetworkInterfaces[0].Subnetwork
+	return
```
Comment (abhinavdahiya) [inline pkg/controller/remotemachineset/gcpactuator.go:489]:
```suggestion
	return remoteMachineProviderSpec.NetworkInterfaces[0].Network, remoteMachineProviderSpec.NetworkInterfaces[0].Subnetwork, nil
```

File: pkg/controller/remotemachineset/gcpactuator.go [hunk 10]
```diff
@@ -451,6 +463,32 @@ func getGCPImageID(masterMachine *machineapi.Machine, scheme *runtime.Scheme, lo
 	return imageID, nil
 }
 
+// Get the network information from existing machines on the remote cluster.
+func getNetwork(remoteMachineSets []machineapi.MachineSet,
```
Comment (abhinavdahiya) [inline pkg/controller/remotemachineset/gcpactuator.go:467]: let's make sure we add unit tests for this function.

File: pkg/controller/remotemachineset/gcpactuator.go [hunk 11]
```diff
@@ -163,8 +173,10 @@ func (a *GCPActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, pool *hi
 	ic := &installertypes.InstallConfig{
 		Platform: installertypes.Platform{
 			GCP: &installertypesgcp.Platform{
-				Region:    cd.Spec.Platform.GCP.Region,
-				ProjectID: a.projectID,
+				Region:        cd.Spec.Platform.GCP.Region,
+				ProjectID:     a.projectID,
+				ComputeSubnet: a.subnet,
+				Network:       a.network,
```
Comment (abhinavdahiya) [inline pkg/controller/remotemachineset/gcpactuator.go:179]: also add unit test that checks the network and subnet match when the actuator is configured with those.

Comment (nimrodshn) [reply pkg/controller/remotemachineset/gcpactuator.go:179]: Will do.

File: pkg/controller/remotemachineset/gcpactuator.go (hunk 10 above)
Comment (nimrodshn) [reply pkg/controller/remotemachineset/gcpactuator.go:468]: Sure.

File: pkg/controller/remotemachineset/gcpactuator.go (hunk 8 above)
Comment (nimrodshn) [reply pkg/controller/remotemachineset/gcpactuator.go]: Done.

File: pkg/controller/remotemachineset/gcpactuator.go (hunk 7 above)
Comment (nimrodshn) [reply pkg/controller/remotemachineset/gcpactuator.go]: Done.

File: pkg/controller/remotemachineset/gcpactuator.go [hunk 12]
```diff
@@ -451,6 +463,33 @@ func getGCPImageID(masterMachine *machineapi.Machine, scheme *runtime.Scheme, lo
...
+		remoteMachineSets[0].Spec.Template.Spec.ProviderSpec.Value,
+		scheme,
+	)
+	if err != nil {
+		logger.WithError(err).Warn("cannot decode GCPMachineProviderSpec from master machine")
+		return "", "", errors.Wrap(err, "cannot decode GCPMachineProviderSpec from master machine")
+	}
+
+	if len(remoteMachineProviderSpec.NetworkInterfaces) == 0 {
+		logger.Warn("remove machine does not have any netowrk interfaces")
```
Comment (abhinavdahiya) [inline pkg/controller/remotemachineset/gcpactuator.go:484]:
```suggestion
		logger.Warn("remote machine do not have any network interfaces")
```

File: pkg/controller/remotemachineset/gcpactuator.go [hunk 13]
```diff
@@ -451,6 +463,33 @@ func getGCPImageID(masterMachine *machineapi.Machine, scheme *runtime.Scheme, lo
...
+		return "", "", nil
```
Comment (abhinavdahiya) [inline pkg/controller/remotemachineset/gcpactuator.go:485]: should this be an error? vs currently no op?

File: pkg/controller/remotemachineset/gcpactuator.go [hunk 14]
```diff
@@ -451,6 +463,33 @@ func getGCPImageID(masterMachine *machineapi.Machine, scheme *runtime.Scheme, lo
...
+	if len(remoteMachineSets) == 0 {
+		return "", "", nil
+	}
+
+	remoteMachineProviderSpec, err := decodeGCPMachineProviderSpec(
+		remoteMachineSets[0].Spec.Template.Spec.ProviderSpec.Value,
+		scheme,
+	)
+	if err != nil {
+		logger.WithError(err).Warn("cannot decode GCPMachineProviderSpec from master machine")
```
Comment (abhinavdahiya) [inline pkg/controller/remotemachineset/gcpactuator.go:479]:
```suggestion
		logger.WithError(err).Warn("cannot decode GCPMachineProviderSpec from remote machinesets")
```

File: pkg/controller/remotemachineset/gcpactuator.go [hunk 15]
```diff
@@ -451,6 +463,33 @@ func getGCPImageID(masterMachine *machineapi.Machine, scheme *runtime.Scheme, lo
...
+		return "", "", errors.Wrap(err, "cannot decode GCPMachineProviderSpec from master machine")
```
Comment (abhinavdahiya) [inline pkg/controller/remotemachineset/gcpactuator.go:480]:
```suggestion
		return "", "", errors.Wrap(err, "cannot decode GCPMachineProviderSpec from remote machinesets")
```

**Review by abhinavdahiya — COMMENTED** [review]
just a few cosmetic changes and one question

File: pkg/controller/remotemachineset/gcpactuator.go (hunk 13 above)
Comment (nimrodshn) [reply pkg/controller/remotemachineset/gcpactuator.go:485]: Probably an error.

Comment (nimrodshn) [reply pkg/controller/remotemachineset/gcpactuator.go]: Fixed.

**Comment by dgoodwin** [comment]
i will force merge as soon as CI is clean

**Pull request merged**